### DISCUSSION TITLE
Way to intercept link clicks when rendering markdown?

### DISCUSSION THREAD

**Discussion by bradparks** [Textualize/rich#2021] [Q&A] [answered]
I'd like to intercept all link clicks in markdown, and do something different with them internally.

So for example, if markdown was rendered for a link like so:

https://someWebsite.com

and someone clicked on that link, then the link wouldn't be opened in the browser, but would fire an event in my code.

Thanks for any suggestions!

**Comment by willmcgugan** [comment, marked as answer]
Afraid not. It's the terminal itself that decides what to do when the link is clicked. AFAIK there is no way to intercept those.

**Reply by bradparks** [reply]
Thanks for the quick response - So is there a way to get an event when text is clicked in a Syntax object, if it's not a link, or is a link the terminal doesn't know how to use (e.g. unknown protocol?) 

My goal is to have parts of the UI of what I'm building be just simple markdown files, where people can code actions into them, and then have the links be clickable, and fire off callbacks in the code.

Worst case, I can make it be a table generated from json or csv or something like that, but some renderable markup (doesn't have to be markdown) would be great of course

**Reply by willmcgugan** [reply]
Not with Rich. You would have to run a full TUI. You might want to look in to Textual, but be aware it is currently under active development.

**Reply by bradparks** [reply]
that looks super cool - so with Texutual, what would the 2 line exectuvive summary of the approach be? Not asking for code, just broad strokes on where to start

and to be clear, my intent is to have a simple markup anyone can use, that renders links that I can fire off callback in code

**Reply by willmcgugan** [reply]
You can attach a link to text with this syntax `"Hello [@click=hello]World[/]"` Which would call an action:

```python
def action_hello():
    # World was clicked
```

**Reply by bradparks** [reply]
awesome - thanks - much appreciated!

**Reply by bradparks** [reply]
So I tried this, but don't seem to be able to get it to work - here's a sample that renders some text, and should make the "World" clickable, I believe, but the "action_hello" method is never called. I've tried it with a `self` param and no `self` self param, but seem to be missing something.

I also bound the "b" key to do pretty much the same thing, so I could see it work
It should change the background color, and the subtitle of the app.
Thanks for any pointers!

``` python
import os
import sys
from rich.console import RenderableType
from rich.panel import Panel
from rich.text import Text
from textual.app import App
from textual.widgets import Header, Footer, ScrollView
from textual.widgets import Placeholder

class MyApp(App):

    async def on_load(self) -> None:
        await self.bind("b", "color('blue')")

    async def on_mount(self) -> None:
        await self.view.dock(Header(), size=5, edge="top")
        await self.view.dock(Footer(), edge="bottom")
        await self.view.dock(ScrollView(Panel("Hello [@click=hello]World[/], [red]more[/] text")), edge="top")

    async def action_color(self, color:str) -> None:
        self.app.sub_title = "KEYBOARD"
        self.background = f"on {color}"

    async def action_hello(self) -> None:
        self.app.sub_title = "CLICKED"
        self.background = "on red"

MyApp.run(title="Test click", log="textual.log")


```

**Reply by bradparks** [reply]
I will take this over to the textual discussions, as that's probably where it's best asked of course 

https://github.com/Textualize/textual/discussions/323
